data_structures/raster_source.hpp

```diff
@@ -0,0 +1,185 @@
+/*
+
+Copyright (c) 2015, Project OSRM contributors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+Redistributions of source code must retain the above copyright notice, this list
+of conditions and the following disclaimer.
+Redistributions in binary form must reproduce the above copyright notice, this
+list of conditions and the following disclaimer in the documentation and/or
+other materials provided with the distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef RASTER_SOURCE_HPP
+#define RASTER_SOURCE_HPP
+
+#include "../util/osrm_exception.hpp"
+
+#include <boost/filesystem.hpp>
+#include <boost/filesystem/fstream.hpp>
+#include <boost/spirit/include/qi_int.hpp>
+#include <boost/spirit/include/qi.hpp>
+#include <boost/algorithm/string/trim.hpp>
+#include <boost/assert.hpp>
+
+#include <unordered_map>
+#include <vector>
+#include <string>
+#include <limits>
+#include <cstdint>
+#include <sstream>
+#include <iterator>
+#include <iostream>
+#include <ios>
+
+/**
+    \brief Small wrapper around raster source queries to optionally provide results
+    gracefully, depending on source bounds
+*/
+struct RasterDatum
+{
+    static std::int32_t get_invalid() { return std::numeric_limits<std::int32_t>::max(); }
+
+    std::int32_t datum = get_invalid();
+
+    RasterDatum() = default;
+
+    RasterDatum(std::int32_t _datum) : datum(_datum){};
+};
+
+class RasterGrid
+{
+  public:
+    RasterGrid(const std::string &filepath, std::size_t _xdim, std::size_t _ydim)
+    {
+        xdim = _xdim;
+        ydim = _ydim;
+        _data.reserve(ydim * xdim);
+
+        boost::filesystem::ifstream stream(boost::filesystem::path{filepath});
```

Comment:
@daniel-j-h — just spent a little more time digging around. Since Lua's types are fairly primitive, the API that in the bindings (`SourceContainer::loadRasterSource`) needs to keep a `std::string` as its function signature. Moreover, the `std::unordered_map LoadedSourcePaths` template can't take a `boost::filesystem::path`, so paths need to be a regular `std::string` for that as well.

So: no, it's impossible to use _only_ `boost::filesystem::path`s throughout this data structure — at some point it needs to construct a `boost::filesystem::path` from an original `std::string`. I changed this slightly in https://github.com/Project-OSRM/osrm-backend/commit/96ff3f17dabaed1ad71ca4fa7dd696fae9b498af so that it constructs a `boost` path only once and a `RasterGrid` constructor takes a boost path, but that's kind of the best we can do here, I think :\

(Also, I'd like to keep the `exists` check and the `stream` check both — an `exists` error implies a user-created problem where the path was specified incorrectly in a profile and therefore belongs in the loading API, whereas a `stream` error speaks maybe to a more system-level error if a path does indeed exist, I think. Just my intuition.)